### PR TITLE
pkg/sink(ticdc): add oauth support for sarama Kafka sink

### PR DESCRIPTION
**Pull request by hi-rustin**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflow/issues/8865

### What is changed and how it works?
add oauth support for sarama Kafka sink. 

Added:
- SASLOAuthClientID  (required)
- SASLOAuthClientSecret  (required)
- SASLOAuthTokenURL (required)
- SASLOAuthScopes 
- SASLOAuthGrantType (default: client_credentials)

#### Tests 

 - [Manual test](https://github.com/pingcap/tiflow/pull/8938#issuecomment-1556648439)

##### Will it cause performance regression or break compatibility?
No
##### Do you need to update user documentation, design documentation or monitoring documentation?
No
### Release note 

```release-note
Add OAuth support for Sarama Kafka sink
```

### CODE REVIEW

**Review by hi-rustin — COMMENTED** [review]
- [x] Add tests for token provider
- [ ] Add test for applying parameters
- [ ] Add test for kafka sink v2

**Comment by hi-rustin** [comment]
Tested locally:
1. start the tidb cluster: `tiup playground v6.5.0 --tiflash 0`
2. start the Kafka cluster: https://github.com/jairsjunior/kafka-playground
3. create a changefeed with config: `./cdc cli changef
eed create --sink-uri="kafka://127.0.0.1:9092/topic1?max-message-bytes=1048576&protocol=open-protocol&kafka-version=2.0.0&replication-factor=1" --config=changefeed.toml`
```toml
[sink]
dispatchers = [
    {matcher = ['test.*', 'test2.*'], dispatcher = "ts"}
]
kafka-config.sasl-mechanism = "OAUTHBEARER"
kafka-config.sasl-oauth-client-id = "producer-kafka"
kafka-config.sasl-oauth-client-secret = "cHJvZHVjZXIta2Fma2E="
kafka-config.sasl-oauth-token-url = "http://127.0.0.1:4444/oauth2/token"
kafka-config.sasl-oauth-scopes = ["producer.kafka"]
kafka-config.sasl-oauth-grant-type = "client_credentials"
```

```log
➜  bin git:(rustin-patch-oauth) ✗ ./cdc cli changefeed create --sink-uri="kafka://127.0.0.1:9092/topic1?max-message-bytes=1048576&protocol=open-protocol&kafka-version=2.0.0&replication-factor=1" --config=changefeed.toml
Create changefeed successfully!
ID: 6ef25092-9ac6-426f-ac1e-76b49774f15b
Info: {"upstream_id":7235897581026360950,"namespace":"default","id":"6ef25092-9ac6-426f-ac1e-76b49774f15b","sink_uri":"kafka://127.0.0.1:9092/topic1?max-message-bytes=1048576\u0026protocol=open-protocol\u0026kafka-version=2.0.0\u0026replication-factor=1","create_time":"2023-05-22T15:00:50.226503677+08:00","start_ts":441644181007106051,"config":{"memory_quota":1073741824,"case_sensitive":true,"enable_old_value":true,"force_replicate":false,"ignore_ineligible_table":false,"check_gc_safe_point":true,"enable_sync_point":false,"bdr_mode":false,"sync_point_interval":600000000000,"sync_point_retention":86400000000000,"filter":{"rules":["*.*"],"event_filters":null},"mounter":{"worker_num":16},"sink":{"protocol":"open-protocol","schema_registry":"","csv":{"delimiter":",","quote":"\"","null":"\\N","include_commit_ts":false},"dispatchers":[{"matcher":["test.*","test2.*"],"partition":"","topic":""}],"column_selectors":null,"transaction_atomicity":"","encoder_concurrency":16,"terminator":"\r\n","date_separator":"none","enable_partition_separator":true,"file_index_width":0,"enable_kafka_sink_v2":false,"only_output_updated_columns":null,"kafka_config":{"sasl_mechanism":"OAUTHBEARER","sasl_oauth_client_id":"producer-kafka","sasl_oauth_client_secret":"cHJvZHVjZXIta2Fma2E=","sasl_oauth_token_url":"http://127.0.0.1:4444/oauth2/token","sasl_oauth_scopes":["producer.kafka"],"sasl_oauth_grant_type":"client_credentials"}},"consistent":{"level":"none","max_log_size":64,"flush_interval":2000,"storage":"","use_file_backend":false},"scheduler":{"enable_table_across_nodes":false,"region_threshold":100000,"write_key_threshold":0},"integrity":{"integrity_check_level":"none","corruption_handle_level":"warn"}},"state":"normal","creator_version":"v7.1.0-master-dirty","resolved_ts":441644181007106051,"checkpoint_ts":441644181007106051,"checkpoint_time":"2023-05-22 15:00:49.667"}
```
4. check the kafka messages: 
```log
{ topic: 'topic1',
  value: '\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000',
  offset: 229,
  partition: 0,
  highWaterOffset: 230,
  key:
   '\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0001\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u001f{"ts":441644203708776450,"t":3}',
  timestamp: 2023-05-22T07:02:16.347Z }
```

**Review by sdojjy — APPROVED** [review]

**Comment by hi-rustin** [comment]
KOP test:
1. Create an API on Auth0
2. Find its test application in Applications
	1. Use domain as the issuer URL
	2. Get the client ID and client secret
	3.  Get the signing secret from the Token Settings of the API
3. Start the pulsar cluster
```yaml
version: '3.5'

services:
  standalone:
    container_name: standalone
    hostname: localhost
    image: streamnative/sn-pulsar:2.10.3.7
    command: >
      bash /waitForOAuth.sh
    environment:
      allowAutoTopicCreationType: partitioned
      brokerDeleteInactiveTopicsEnabled: "false"
      PULSAR_PREFIX_saslAllowedMechanisms: OAUTHBEARER
      PULSAR_PREFIX_kopOauth2AuthenticateCallbackHandler: io.streamnative.pulsar.handlers.kop.security.oauth.OauthValidatorCallbackHandler
      PULSAR_PREFIX_kopOauth2ConfigFile: /conf/kop-handler.properties
      PULSAR_PREFIX_messagingProtocols: kafka
      PULSAR_PREFIX_kafkaListeners: SASL_PLAINTEXT://0.0.0.0:9092
      PULSAR_PREFIX_kafkaAdvertisedListeners: SASL_PLAINTEXT://xxxx:19092
      PULSAR_PREFIX_brokerEntryMetadataInterceptors: org.apache.pulsar.common.intercept.AppendIndexMetadataInterceptor
      PULSAR_PREFIX_authenticationEnabled: true
      PULSAR_PREFIX_authenticationProviders: org.apache.pulsar.broker.authentication.AuthenticationProviderToken
      PULSAR_PREFIX_brokerClientAuthenticationPlugin: org.apache.pulsar.client.impl.auth.oauth2.AuthenticationOAuth2
      PULSAR_PREFIX_brokerClientAuthenticationParameters: "{\"type\":\"client_credentials\",\"privateKey\":\"/conf/client_credentials.json\",\"issuerUrl\":\"https://dev-xxx.us.auth0.com\", \"audience\": \"pulsar\"}"
      PULSAR_PREFIX_tokenSecretKey: "data:;base64,xxxxxx="
      LOGGING_LEVEL: debug
    ports:
      - 6650:6650
      - 8080:8080
      - 19092:9092
    volumes:
      - ./client_credentials.json:/conf/client_credentials.json
      - ./kop-handler.properties:/conf/kop-handler.properties
      - ./waitForOAuth.sh:/waitForOAuth.sh
```
client_credentials.json:
```json
{
  "type": "client_credentials",
  "client_id": "h2IA1jxxjxFZBgpX6z",
  "client_secret": "-BHtxxxxGW6bJwfihRO9YN2ADUl",
  "issuer_url": "https://dxxx44.us.auth0.com",
  "audience": "pulsar"
}
kop-handler.properties:
```properties
oauth.validate.method=token
```
waitForOAuth.sh:
```sh
bin/apply-config-from-env.py conf/standalone.conf && exec bin/pulsar standalone
```
4. start the tidb cluster: `tiup playground v6.5.0 --tiflash 0`
5. start the ticdc cluster: `./cdc server --log-level=debug  > test.log`
6. create a chanegfeed with this config: `./cdc cli changefeed create --sink-uri="kafka://10.2.7.27:19092/testdsada?max-message-bytes=1048576&protocol=open-protocol&kafka-version=2.0.0&replication-factor=1" --config=./changefeed.toml`
```toml
[sink]
kafka-config.sasl-mechanism = "OAUTHBEARER"
kafka-config.sasl-oauth-client-id = "xxx"
kafka-config.sasl-oauth-client-secret = "xxxx=="
kafka-config.sasl-oauth-token-url = "https://dxxxh/token"
kafka-config.sasl-oauth-grant-type = "client_credentials"
kafka-config.sasl-oauth-audience = "pulsar"
```

7. write some data:
```sql
mysql> use test;
Database changed
mysql> CREATE TABLE example1 (     id INT PRIMARY KEY,     b  INT );
Query OK, 0 rows affected (0.12 sec)

mysql> INSERT INTO example1 (id, b) VALUES (2, 2);
Query OK, 1 row affected (0.01 sec)
```
8. check the changefeed lag:
```log
 bin git:(rustin-patch-oauth) ✗ ./cdc cli changefeed list
[
  {
    "id": "e3b3b577-2197-4cfc-9b42-e0a1775abe44",
    "namespace": "default",
    "summary": {
      "state": "normal",
      "tso": 441689899176558596,
      "checkpoint": "2023-05-24 15:27:30.643",
      "error": null
    }
  }
]
```

File: pkg/sink/kafka/options.go [hunk 1]
```diff
@@ -397,57 +400,119 @@ func (o *Options) applyTLS(params *urlConfig) error {
...
+		if replicaConfig.Sink.KafkaConfig.SASLOAuthTokenURL != nil {
+			tokenURL := *replicaConfig.Sink.KafkaConfig.SASLOAuthTokenURL
+			if tokenURL == "" {
+				return cerror.ErrKafkaInvalidConfig.GenWithStack(
+					"OAuth2 token URL cannot be empty")
+			}
+			o.SASL.OAuth2.TokenURL = tokenURL
+		}
+
+		if o.SASL.OAuth2.IsEnable() {
```
[the remark this answers is not in the document]
Comment (hi-rustin) [reply pkg/sink/kafka/options.go:492]: Yes, we have to pass the required parameters first.

**Review by nongfushanquan — APPROVED** [review]

**Pull request merged**